site/blocks/StickyNav.tsx

```diff
@@ -0,0 +1,202 @@
+    x: number
+    y: number
+} {
+    let y = 0
+    let x = 0
+    while (true) {
+        x += element.offsetLeft
+        y += element.offsetTop
+        if (!element.offsetParent) {
+            break
+        }
+        element = element.offsetParent as HTMLElement
+    }
+    return { x, y }
+}
+
+const STICKY_NAV_HEIGHT = 54
+const HEADING_SCROLL_MARGIN = 16
+
+interface StickyNavProps {
+    links: { text: string; target: string }[]
+}
+class StickyNav extends React.Component<
+    StickyNavProps,
+    {
+        headingPositions: { id: string; yPosition: number }[]
+        currentHeadingIndex?: number
+        links: { text: string; target: string }[]
+    }
+> {
+    ulRef = createRef<HTMLUListElement>()
+
+    state = {
+        currentHeadingIndex: undefined,
+        headingPositions: [] as { id: string; yPosition: number }[],
+        links: this.props.links,
+    }
+
+    handleResize = throttle(() => {
+        this.setHeadingPositions()
+    }, 50)
+
+    handleScroll = throttle(() => {
+        this.setCurrentHeading()
+    }, 50)
+
+    // Make sure that each item in the nav actually points to a heading in the page
+    filterValidLinks() {
+        const validLinks = this.state.links.filter(
+            (link) =>
+                link.target.startsWith("#") &&
+                document.querySelector(link.target)
+        )
+
+        this.setState({
+            links: validLinks,
+        })
+    }
+
+    setCurrentHeading() {
+        const { scrollY } = window
+        let currentHeadingIndex = 0
+        for (let i = 0; i < this.state.headingPositions.length; i++) {
+            const heading = this.state.headingPositions[i]
+            const target =
+                heading.yPosition -
+                STICKY_NAV_HEIGHT -
+                HEADING_SCROLL_MARGIN -
+                // Additional clearance for height of actual element
+                32
+            if (scrollY <= target) {
+                break
+            } else {
+                currentHeadingIndex = i
+            }
+        }
+
+        const current = this.ulRef.current
+        if (current) {
+            const parent = current.parentElement
+            if (parent) {
+                const hasScrollbar = parent.scrollWidth < current.scrollWidth
+                if (hasScrollbar) {
+                    const listItem = current.children[
+                        currentHeadingIndex
+                    ] as HTMLElement
+                    const containerPadding = 32
+                    current.scrollTo({
+                        behavior: "smooth",
+                        left: listItem.offsetLeft - containerPadding,
+                    })
+                }
+            }
+        }
+
+        this.setState({
+            currentHeadingIndex,
+        })
+    }
+
+    setHeadingPositions() {
+        const headingPositions: { id: string; yPosition: number }[] = []
+
+        this.state.links.forEach(({ target }, i) => {
```

Comment:
I _still_ think this should be a `.map()` here :)

Reply:
Ordinarily I'd agree with you but it needs to be `forEach` so we can access `headingPositions[i - 1]` in the else statement. 🙂

Reply:
Ah, you're right of course!